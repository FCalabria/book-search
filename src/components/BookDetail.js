import React, {useState} from 'react';
import {LanguageConsumer} from '../language-context'

function BookDetail(props) {
  const [imageId, setImageId] = useState(props.isbn[0])
  const [imageIdIndex, setImageIdIndex] = useState(0)

  function onImageError() {
    const newIndex = imageIdIndex + 1
    if (props.isbn[newIndex]) {
      setImageId(props.isbn[newIndex])
      setImageIdIndex(newIndex)
    } else {
      setImageId('')
    }
  }
    return (
    <LanguageConsumer>
    {language => (
        <div className="flex flex-wrap md:flex-no-wrap px-3 py-4 odd:bg-teal-300">
          <div className="w-20 sm:w-32 lg:w-40 mr-2 flex-grow-0 flex-shrink-0 bg-pink-200">
            <img
              src={imageId
                ? `http://covers.openlibrary.org/b/isbn/${imageId}-M.jpg?default=false`
                : 'https://dummyimage.com/130x170/fed7e2/fed7e2.png'
              }
              alt=""
              onError={onImageError}/>
          </div>
          <div className="flex flex-1 flex-wrap content-center mr-2">
            <span className="w-full text-xl font-medium">{props.title}</span>
            <span className="italic text-gray-700">{props.author}</span>
            <span className="text-gray-700">&nbsp;-&nbsp;{props.year}</span>
          </div>
          {
            props.hasSubject &&
            <button className="self-center w-full md:w-auto flex-none p-2 mt-3 md:mt-0 ml-auto shadow-lg rounded-lg border-2 border-pink-700 text-pink-700 hover:bg-teal-100 active:bg-teal-100 focus:bg-teal-100" onClick={props.onSearchSimilar}>{language.similar}</button>
          }
      </div>
      )}
    </LanguageConsumer>
    )
}

export default BookDetail