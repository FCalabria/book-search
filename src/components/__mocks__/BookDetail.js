import React from 'react'
export default (props) => {
  return <div>
    {props.author}
    <button onClick={props.onSearchSimilar}>similar button mock</button>
  </div>;
}