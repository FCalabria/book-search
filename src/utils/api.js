async function apiFetch (path, fetchOptions) {
  try {
    const result = await fetch(`http://openlibrary.org/${path}`, fetchOptions)
    if (result.ok) {
      const response = await result.json()
      return Promise.resolve(response)
    } else {
      return Promise.reject(result)
    }
  } catch (error) {
    console.log(`Fetch error ${path}`, error)
    return Promise.reject(new Error('Network error: check your connection and try again'))
  }
}

export default apiFetch