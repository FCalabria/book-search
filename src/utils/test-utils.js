import React from 'react'
import { render } from '@testing-library/react'
import {LanguageProvider, languages} from '../language-context'


const AllTheProviders = ({ children }) => {
  return (
    <LanguageProvider value={{
      language: languages.ES,
      keys: Object.keys(languages),
      selected: 'ES',
      toggleLanguage: () => {}
    }}>
      {children}
    </LanguageProvider>
  )
}

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
