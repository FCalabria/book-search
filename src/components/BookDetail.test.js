import React from 'react';
import { render } from '../utils/test-utils';
import BookDetail from './BookDetail';

describe('BookDetail', () => {
  const defaultProps = {
    isbn: ['fake-isbn'],
    title: 'book title',
    author: 'someone',
    year: '1999'
  }
  let baseContainer
  beforeEach(() => {
    baseContainer = render(<BookDetail {...defaultProps} />).container;
  })
  test('should render correctly when there is no error', () => {
    expect(baseContainer.firstChild).toMatchSnapshot();
  });
  test('should render the "similar" button when it has a subjet', () => {
    const { container } = render(<BookDetail {...defaultProps} hasSubject={true} />);
    expect(baseContainer.firstChild).toMatchDiffSnapshot(container.firstChild);
  })
})
