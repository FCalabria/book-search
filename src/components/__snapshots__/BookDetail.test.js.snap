// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetail should render correctly when there is no error 1`] = `
<div
  class="flex flex-wrap md:flex-no-wrap px-3 py-4 odd:bg-teal-300"
>
  <div
    class="w-20 sm:w-32 lg:w-40 mr-2 flex-grow-0 flex-shrink-0 bg-pink-200"
  >
    <img
      alt=""
      src="http://covers.openlibrary.org/b/isbn/fake-isbn-M.jpg?default=false"
    />
  </div>
  <div
    class="flex flex-1 flex-wrap content-center mr-2"
  >
    <span
      class="w-full text-xl font-medium"
    >
      book title
    </span>
    <span
      class="italic text-gray-700"
    >
      someone
    </span>
    <span
      class="text-gray-700"
    >
       - 
      1999
    </span>
  </div>
</div>
`;

exports[`BookDetail should render the "similar" button when it has a subjet 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -27,6 +27,11 @@
      >
         - 
        1999
      </span>
    </div>
+   <button
+     class=\\"self-center w-full md:w-auto flex-none p-2 mt-3 md:mt-0 ml-auto shadow-lg rounded-lg border-2 border-pink-700 text-pink-700 hover:bg-teal-100 active:bg-teal-100 focus:bg-teal-100\\"
+   >
+     Buscar parecido
+   </button>
  </div>"
`;
